subsys/lorawan/services/frag_decoder_lowmem.c

```diff
@@ -0,0 +1,411 @@
+/*
+ * Copyright (c) 2024 A Labs GmbH
+ * Copyright (c) 2024 tado GmbH
+ * Copyright (c) 2022 Jiapeng Li
+ *
+ * Based on: https://github.com/JiapengLi/LoRaWANFragmentedDataBlockTransportAlgorithm
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "frag_decoder_lowmem.h"
+#include "frag_flash.h"
+
+#include <zephyr/logging/log.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/sys/bitarray.h>
+
+LOG_MODULE_REGISTER(lorawan_frag_dec, CONFIG_LORAWAN_SERVICES_LOG_LEVEL);
+
+SYS_BITARRAY_DEFINE_STATIC(lost_frames, FRAG_MAX_NB);
+SYS_BITARRAY_DEFINE_STATIC(lost_frames_matrix, (FRAG_TOLERANCE * (FRAG_TOLERANCE + 1) / 2));
+SYS_BITARRAY_DEFINE_STATIC(matched_lost_frm_bm0, FRAG_TOLERANCE);
+SYS_BITARRAY_DEFINE_STATIC(matched_lost_frm_bm1, FRAG_TOLERANCE);
+SYS_BITARRAY_DEFINE_STATIC(matrix_line_bm, FRAG_MAX_NB);
+
+static inline size_t matrix_location_to_index(size_t x, size_t y, size_t m);
+
+static bool triangular_matrix_get_entry(struct sys_bitarray *m2tbm, size_t x, size_t y, size_t m);
+static void triangular_matrix_set_entry(struct sys_bitarray *m2tbm, size_t x, size_t y, size_t m);
+static void triangular_matrix_clear_entry(struct sys_bitarray *m2tbm, size_t x, size_t y, size_t m);
+
+static inline bool bit_get(struct sys_bitarray *bitmap, size_t index);
+static inline void bit_set(struct sys_bitarray *bitmap, size_t index);
+static inline void bit_clear(struct sys_bitarray *bitmap, size_t index);
+static inline size_t bit_count_ones(struct sys_bitarray *bitmap, size_t index);
+static inline void bit_xor(struct sys_bitarray *des, struct sys_bitarray *src, size_t size);
+static inline void bit_clear_all(struct sys_bitarray *bitmap, size_t size);
+
+/**
+ * Generate a 23bit Pseudorandom Binary Sequence (PRBS)
+ *
+ * @param seed Seed input value
+ *
+ * @returns Pseudorandom output value
+ */
+static int32_t prbs23(int32_t seed);
```

Comment:
The parameter is not a seed value in the traditional sense, I think `previous` is more accurate.
A PRBS generates the next value from the previous.